File: examples/ThunderAuto/java/src/main/java/frc/robot/SwerveModule.java
package frc.robot;

// Modified version of SwerveModule class from WPILib SwerveBot example.

import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.controller.ProfiledPIDController;
import edu.wpi.first.math.controller.SimpleMotorFeedforward;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.kinematics.SwerveModulePosition;
import edu.wpi.first.math.kinematics.SwerveModuleState;
import edu.wpi.first.math.trajectory.TrapezoidProfile;
import edu.wpi.first.wpilibj.Encoder;
import edu.wpi.first.wpilibj.RobotBase;
import edu.wpi.first.wpilibj.motorcontrol.PWMSparkMax;
import edu.wpi.first.wpilibj.simulation.EncoderSim;
import static frc.robot.Constants.*;
import static frc.robot.Constants.DriveConstants.*;

public class SwerveModule {
  private final PWMSparkMax driveMotor;
  private final PWMSparkMax turningMotor;

  private final Encoder driveEncoder;
  private final Encoder turningEncoder;

  private final EncoderSim driveEncoderSim;
  private final EncoderSim turningEncoderSim;

  // Gains are for example purposes only - must be determined for your own robot!
  private final PIDController drivePIDController = new PIDController(1, 0, 0);

  // Gains are for example purposes only - must be determined for your own robot!
  private final ProfiledPIDController turningPIDController = new ProfiledPIDController(
      1,
      0,
      0,
      new TrapezoidProfile.Constraints(kModuleMaxTurningVelocity, kModuleMaxTurningAcceleration));

  // Gains are for example purposes only - must be determined for your own robot!
  private final SimpleMotorFeedforward driveFeedforward = new SimpleMotorFeedforward(1, 3);
  private final SimpleMotorFeedforward turnFeedforward = new SimpleMotorFeedforward(1, 0.5);

  /**
   * Constructs a SwerveModule with a drive motor, turning motor, drive encoder
   * and turning encoder.
   *
   * @param driveMotorChannel      PWM output for the drive motor.
   * @param turningMotorChannel    PWM output for the turning motor.
   * @param driveEncoderChannelA   DIO input for the drive encoder channel A
   * @param driveEncoderChannelB   DIO input for the drive encoder channel B
   * @param turningEncoderChannelA DIO input for the turning encoder channel A
   * @param turningEncoderChannelB DIO input for the turning encoder channel B
   */
  public SwerveModule(
      int driveMotorChannel,
      int turningMotorChannel,
      int driveEncoderChannelA,
      int driveEncoderChannelB,
      int turningEncoderChannelA,
      int turningEncoderChannelB) {
    driveMotor = new PWMSparkMax(driveMotorChannel);
    turningMotor = new PWMSparkMax(turningMotorChannel);

    driveEncoder = new Encoder(driveEncoderChannelA, driveEncoderChannelB);
    turningEncoder = new Encoder(turningEncoderChannelA, turningEncoderChannelB);

    // Set the distance per pulse for the drive encoder. We can simply use the
    // distance traveled for one rotation of the wheel divided by the encoder
    // resolution.
    driveEncoder.setDistancePerPulse(
        2 * Math.PI * kWheelRadius / kEncoderResolution);

    // Set the distance (in this case, angle) in radians per pulse for the turning
    // encoder.
    // This is the the angle through an entire rotation (2 * pi) divided by the
    // encoder resolution.
    turningEncoder.setDistancePerPulse(2 * Math.PI / kEncoderResolution);

    driveEncoderSim = new EncoderSim(driveEncoder);
    turningEncoderSim = new EncoderSim(turningEncoder);

    // Limit the PID Controller's input range between -pi and pi and set the input
    // to be continuous.
    turningPIDController.enableContinuousInput(-Math.PI, Math.PI);
  }

  /**
   * Returns the current state of the module.
   *
   * @return The current state of the module.
   */
  public SwerveModuleState getState() {
    return new SwerveModuleState(
        driveEncoder.getRate(), new Rotation2d(turningEncoder.getDistance()));
  }

  /**
   * Returns the current position of the module.
   *
   * @return The current position of the module.
   */
  public SwerveModulePosition getPosition() {
    return new SwerveModulePosition(
        driveEncoder.getDistance(), new Rotation2d(turningEncoder.getDistance()));
  }

  /**
   * Sets the desired state for the module.
   *
   * @param desiredState Desired state with speed and angle.
   */
  public void setDesiredState(SwerveModuleState desiredState) {
    if (RobotBase.isSimulation()) {
      setDesiredStateSim(desiredState);
    } else {
      setDesiredStateReal(desiredState);
    }
  }

  private void setDesiredStateReal(SwerveModuleState desiredState) {
    Rotation2d encoderRotation = new Rotation2d(turningEncoder.getDistance());

    // Optimize the reference state to avoid spinning further than 90 degrees
    desiredState.optimize(encoderRotation);

    // Scale speed by cosine of angle error. This scales down movement perpendicular
    // to the desired direction of travel that can occur when modules change
    // directions. This results in smoother driving.
    desiredState.cosineScale(encoderRotation);

    // Calculate the drive output from the drive PID controller.

    final double driveOutput = drivePIDController.calculate(driveEncoder.getRate(),
        desiredState.speedMetersPerSecond);

    final double driveFeedforwardOutput = driveFeedforward.calculate(desiredState.speedMetersPerSecond);

    // Calculate the turning motor output from the turning PID controller.
    final double turnOutput = turningPIDController.calculate(
        turningEncoder.getDistance(), desiredState.angle.getRadians());

    final double turnFeedforwardOutput = turnFeedforward.calculate(turningPIDController.getSetpoint().velocity);

    driveMotor.setVoltage(driveOutput + driveFeedforwardOutput);
    turningMotor.setVoltage(turnOutput + turnFeedforwardOutput);
  }

  private void setDesiredStateSim(SwerveModuleState desiredState) {
    Rotation2d encoderRotation = new Rotation2d(turningEncoder.getDistance());

    // Optimize the reference state to avoid spinning further than 90 degrees
    desiredState.optimize(encoderRotation);

    // Scale speed by cosine of angle error. This scales down movement perpendicular
    // to the desired direction of travel that can occur when modules change
    // directions. This results in smoother driving.
    desiredState.cosineScale(encoderRotation);

    // Simulate setting the speed and angle by directly setting the encoder values.
    driveEncoderSim.setRate(desiredState.speedMetersPerSecond);
    driveEncoderSim.setDistance(
        driveEncoder.getDistance() + desiredState.speedMetersPerSecond * kPeriodSeconds);
    turningEncoderSim.setDistance(desiredState.angle.getRadians());
  }
}

File: examples/ThunderAuto/cpp/src/main/cpp/SwerveModule.cpp
#include "SwerveModule.h"

#include <numbers>

#include <frc/geometry/Rotation2d.h>
#include <frc/RobotBase.h>

SwerveModule::SwerveModule(const int driveMotorChannel,
                           const int turningMotorChannel,
                           const int driveEncoderChannelA,
                           const int driveEncoderChannelB,
                           const int turningEncoderChannelA,
                           const int turningEncoderChannelB)
    : driveMotor(driveMotorChannel),
      turningMotor(turningMotorChannel),
      driveEncoder(driveEncoderChannelA, driveEncoderChannelB),
      turningEncoder(turningEncoderChannelA, turningEncoderChannelB) {
  // Set the distance per pulse for the drive encoder. We can simply use the
  // distance traveled for one rotation of the wheel divided by the encoder
  // resolution.
  driveEncoder.SetDistancePerPulse(2 * std::numbers::pi * DriveConstants::kWheelRadius.value() /
                                   DriveConstants::kEncoderResolution);

  // Set the distance (in this case, angle) per pulse for the turning encoder.
  // This is the the angle through an entire rotation (2 * std::numbers::pi)
  // divided by the encoder resolution.
  turningEncoder.SetDistancePerPulse(2 * std::numbers::pi / DriveConstants::kEncoderResolution);

  // Limit the PID Controller's input range between -pi and pi and set the input
  // to be continuous.
  turningPIDController.EnableContinuousInput(-units::radian_t{std::numbers::pi},
                                             units::radian_t{std::numbers::pi});
}

frc::SwerveModuleState SwerveModule::GetState() const {
  return {units::meters_per_second_t{driveEncoder.GetRate()}, units::radian_t{turningEncoder.GetDistance()}};
}

frc::SwerveModulePosition SwerveModule::GetPosition() const {
  return {units::meter_t{driveEncoder.GetDistance()}, units::radian_t{turningEncoder.GetDistance()}};
}

void SwerveModule::SetDesiredState(frc::SwerveModuleState& referenceState) {
  if (frc::RobotBase::IsSimulation()) {
    SetDesiredStateSim(referenceState);
  } else {
    SetDesiredStateReal(referenceState);
  }
}

void SwerveModule::SetDesiredStateReal(frc::SwerveModuleState& referenceState) {
  frc::Rotation2d encoderRotation{units::radian_t{turningEncoder.GetDistance()}};

  // Optimize the reference state to avoid spinning further than 90 degrees
  referenceState.Optimize(encoderRotation);

  // Scale speed by cosine of angle error. This scales down movement perpendicular to the desired direction of
  // travel that can occur when modules change directions. This results in smoother driving.
  referenceState.CosineScale(encoderRotation);

  // Calculate the drive output from the drive PID controller.
  const units::volt_t driveOutput{
      drivePIDController.Calculate(driveEncoder.GetRate(), referenceState.speed.value())};

  const units::volt_t driveFeedforwardOutput = driveFeedforward.Calculate(referenceState.speed);

  // Calculate the turning motor output from the turning PID controller.
  const units::volt_t turnOutput{turningPIDController.Calculate(units::radian_t{turningEncoder.GetDistance()},
                                                                referenceState.angle.Radians())};

  const units::volt_t turnFeedforwardOutput =
      turnFeedforward.Calculate(turningPIDController.GetSetpoint().velocity);

  // Set the motor outputs.
  driveMotor.SetVoltage(driveOutput + driveFeedforwardOutput);
  turningMotor.SetVoltage(turnOutput + turnFeedforwardOutput);
}

void SwerveModule::SetDesiredStateSim(frc::SwerveModuleState& referenceState) {
  frc::Rotation2d encoderRotation{units::radian_t{turningEncoder.GetDistance()}};

  // Optimize the reference state to avoid spinning further than 90 degrees
  referenceState.Optimize(encoderRotation);

  // Scale speed by cosine of angle error. This scales down movement perpendicular to the desired direction of
  // travel that can occur when modules change directions. This results in smoother driving.
  referenceState.CosineScale(encoderRotation);

  // Simulate setting the speed and angle by directly setting the encoder values.
  driveEncoderSim.SetRate(referenceState.speed.value());
  driveEncoderSim.SetDistance(driveEncoder.GetDistance() +
                              referenceState.speed.value() * kPeriodicUpdateTime.value());
  turningEncoderSim.SetDistance(referenceState.angle.Radians().value());
}
